### ISSUE TITLE
Weight decay types should be weight regularizers

### ISSUE THREAD

**Issue by kparichay** [nnstreamer/nntrainer#456]
Weight decay is just another name for L2 regularization for weights. (There is slight difference on how its applied with learning rate and opitmizers but the concept is same).
nntrainer treats weight decay as generic weight initializer. 
Weight decay needs to be renamed as weight initializer.

**Comment by jijoongmoon** [comment]
I found good explanation about it 
https://towardsdatascience.com/weight-decay-l2-regularization-90a9e17713cd

According to this article, the difference is that muliplication of learning rate, then.
It also mentions that we don't apply regularization for bias. It say that it causes the uderfitting.
I think we didn't. 

It also say that L2 regularization is not good for adam, but good for sgd. If we using weight decay then it is good for adam and sgd both.
This article also mentioned kind of regularization.
L2 Regularization, L1 Regularization,  Dropout, Early Stopping, Data Augmentation.

I think we have to consider it before we fix the name. I think we have to support by Dropout. ( keras provides this as an layer )
and do you think we have to support weight decay ? ( I guess we do. Cause weight decay good for adam and sgd both like article said )

**Comment by kparichay** [comment]
How about we have two options - 
- Weight Regularizer - This can support L1, L2, etc
- Bias Regularizer - Same support

Add weight decay as an option to the optimizer - If added, this will add weight decay to all the weights in the model (is user specifies weight regularizer and weight_decay, we can give warning and choose one or both).

Different Regularizer (?) : Dropout, Early Stopping, Data Augmentation.

**Comment by jijoongmoon** [comment]
Closed by #503
